akka-actor/src/main/scala/akka/routing/MetricsBasedResizer.scala

```diff
@@ -0,0 +1,242 @@
+/**
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.routing
+
+import java.time.LocalDateTime
+
+import scala.collection.immutable
+import scala.concurrent.forkjoin.ThreadLocalRandom
+import scala.concurrent.duration._
+
+import com.typesafe.config.Config
+
+import akka.actor._
+import akka.util.JavaDurationConverters._
+
+import MetricsBasedResizer._
+
+trait MetricsBasedResizer extends Resizer {
+  /**
+   * Report the messageCount as well as current routees so that the
+   * it can collect metrics.
+   * Caution: this method is not thread safe.
+   *
+   * @param currentRoutees
+   * @param messageCounter
+   */
+  def reportMessageCount(currentRoutees: immutable.IndexedSeq[Routee], messageCounter: Long): Unit
+}
+
+case object MetricsBasedResizer {
+  /**
+   * INTERNAL API
+   */
+  private[routing]type PoolSize = Int
+
+  /**
+   * INTERNAL API
+   */
+  private[routing] case class UnderUtilizationStreak(start: LocalDateTime, highestUtilization: Int)
+
+  /**
+   * INTERNAL API
+   */
+  private[routing] case class ResizeRecord(
+    underutilizationStreak: Option[UnderUtilizationStreak] = None,
+    messageCount: Long = 0,
+    totalQueueLength: Int = 0,
+    checkTime: Long = 0)
+
+  /**
+   * INTERNAL API
+   */
+  private[routing]type PerformanceLog = Map[PoolSize, Duration]
+
+  def apply(resizerCfg: Config): MetricsBasedResizer =
+    DefaultMetricsBasedResizer(
+      lowerBound = resizerCfg.getInt("lower-bound"),
+      upperBound = resizerCfg.getInt("upper-bound"),
+      chanceOfScalingDownWhenFull = resizerCfg.getDouble("chance-of-ramping-down-when-full"),
+      actionInterval = resizerCfg.getDuration("action-interval").asScala,
+      downsizeAfterUnderutilizedFor = resizerCfg.getDuration("downsize-after-underutilized-for").asScala,
+      numOfAdjacentSizesToConsiderDuringOptimization = resizerCfg.getInt("optimization-range"),
+      exploreStepSize = resizerCfg.getDouble("explore-step-size"),
+      explorationProbability = resizerCfg.getDouble("chance-of-exploration"),
+      weightOfLatestMetric = resizerCfg.getDouble("weight-of-latest-metric"),
+      downsizeRatio = resizerCfg.getDouble("downsize-ratio"))
+
+}
+
+/**
+ * Implementation of [[Resizer]] that adjust the [[Pool]]
+ * based on performance per size.
+ * It keeps a performance log so it's stateful as well as
+ * a larger memory footprint than the [[DefaultResizer]].
+ * For documentation about the the parameters, see the reference.conf -
+ * akka.actor.deployment.default.metrics-based-resizer
+ *
+ * CAUTION: [[resize]] also changes the inner state of the resizer,
+ * do not call it unnecessarily or even concurrently
+ *
+ *
+ */
+@SerialVersionUID(1L)
+case class DefaultMetricsBasedResizer(
```

Comment:
Some of the logic must be explained here, if I'm a user I don't know what this resizer is doing based on this doc.
We also have to explain it in https://github.com/akka/akka/blob/master/akka-docs/rst/scala/routing.rst#dynamically-resizable-pool (as sub-section), a small walk-through and rationale is enough :-)